paddle/fluid/operators/concat_op.cc

```diff
@@ -80,8 +80,19 @@ class ConcatOp : public framework::OperatorWithKernel {
  protected:
   framework::OpKernelType GetExpectedKernelType(
       const framework::ExecutionContext &ctx) const override {
-    auto input_data_type =
-        framework::GetDataTypeOfVar(ctx.MultiInputVar("X")[0]);
+    auto vars = ctx.MultiInputVar("X");
+    const size_t num_input = vars.size();
+    auto input_data_type = paddle::framework::proto::VarType::Type(0);
+    for (size_t i = 0; i < num_input; ++i) {
+      try {
+        input_data_type = framework::GetDataTypeOfVar(vars[i]);
+        break;
+      } catch (platform::EnforceNotMet exception) {
+        if (i == num_input - 1) {
+          PADDLE_THROW("All Inputs of Concat OP are Empty!");
+        }
+      }
+    }
```

Comment:
The above implementation is a bit complicated.
You can try to fix like the following:
```
auto inputs = ctx.MultiInputVar("X");
for(auto* input : inputs){
  if(input->IsInitialized()){
    ....
  }
}
```